### PR TITLE
feat: Implement bot detection categories

### PR DESCRIPTION
**Pull request by blaine-arcjet**
This is mostly just codegen and a little bit of type updates.

I'm planning to add a new example for categories, with the additional demonstration of filtering items from a category.

### CODE REVIEW

**Comment by blaine-arcjet** [comment]
@davidmytton I've added an example for the bot categories. Please try it out! I also included usage and details about your filtering request in the example.

**Comment by blaine-arcjet** [comment]
Note to self: add this to dependabot and CI

File: examples/nextjs-14-bot-categories/app/api/arcjet/route.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    // WARNING: This is illustrative! Don't share this metadata with users;
+    // otherwise they may use it to subvert bot detection!
+    headers.set("X-Arcjet-Bot-Allowed",  decision.reason.allowed.join(", "))
+    headers.set("X-Arcjet-Bot-Denied",  decision.reason.denied.join(", "))
+  }
+
+  if (decision.isDenied()) {
+    return NextResponse.json(
+      { error: "You are a Bot!" },
+      { status: 429, headers },
```
Comment (davidmytton) [inline examples/nextjs-14-bot-categories/app/api/arcjet/route.ts:40]:
```suggestion
      { status: 403, headers },
```

**Review by davidmytton — CHANGES_REQUESTED** [review]
I tested the nextjs-14-bot-categories example app, but I don't get the category in the headers or when the decision is logged:

```
< HTTP/1.1 429 Too Many Requests
< vary: RSC, Next-Router-State-Tree, Next-Router-Prefetch
< content-type: application/json
< x-arcjet-bot-allowed:
< x-arcjet-bot-denied: CURL
< Date: Sun, 15 Sep 2024 16:45:40 GMT
< Connection: keep-alive
< Keep-Alive: timeout=5
< Transfer-Encoding: chunked
<
* Connection #0 to host localhost left intact
{"error":"You are a Bot!"}%
```

```
ArcjetDenyDecision {
  id: 'req_01j7v96s53fc7btc6naz433qce',
  results: [
    ArcjetRuleResult {
      ruleId: '',
      ttl: 60,
      state: 'RUN',
      conclusion: 'DENY',
      reason: [ArcjetBotReason]
    }
  ],
  ttl: 60,
  ip: ArcjetIpDetails {
    latitude: undefined,
    longitude: undefined,
    accuracyRadius: undefined,
    timezone: undefined,
    postalCode: undefined,
    city: undefined,
    region: undefined,
    country: undefined,
    countryName: undefined,
    continent: undefined,
    continentName: undefined,
    asn: undefined,
    asnName: undefined,
    asnDomain: undefined,
    asnType: undefined,
    asnCountry: undefined,
    service: undefined
  },
  conclusion: 'DENY',
  reason: ArcjetBotReason { type: 'BOT', allowed: [], denied: [ 'CURL' ] }
```

**Comment by blaine-arcjet** [comment]
@davidmytton did you test against the decide service running locally? Since the example is using an `allow` list, if you tested against the test or prod server, you'd get CURL denied (as categories aren't supported until it is deployed). I also ran into this when testing so the result looks familiar.

**Review by davidmytton — APPROVED** [review]
Yeh I was testing against prod. Just tested against the local branch and all looks good!

**Pull request merged**